packages/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm

```diff
@@ -614,17 +614,27 @@ - (void)handleInputAccessoryDoneButton
 }
 
 - (void)_updateState
+{
+    NSAttributedString *attributedString = _backedTextInputView.attributedText;
+    [self _updateStateWithString:attributedString];
+}
+
+- (void)_updateStateWithString:(NSAttributedString*)attributedString
 {
   if (!_state) {
     return;
   }
-  NSAttributedString *attributedString = _backedTextInputView.attributedText;
   auto data = _state->getData();
   _lastStringStateWasUpdatedWith = attributedString;
   data.attributedStringBox = RCTAttributedStringBoxFromNSAttributedString(attributedString);
   _mostRecentEventCount += _comingFromJS ? 0 : 1;
   data.mostRecentEventCount = _mostRecentEventCount;
-  _state->updateState(std::move(data));
+  const auto &textInputEventEmitter = static_cast<const TextInputEventEmitter &>(*_eventEmitter);
+  // When the textInputDidChange gets called, the text is already updated
+  // in the UI. We execute the state update synchronously so that the layout gets calculated immediately.
+  textInputEventEmitter.experimental_flushSync([state = _state, data = std::move(data)]() mutable {
```

Comment:
the API is not ready for a general adoption. We are still iterating on the design and trying to prove its usefulness in production.

Reply:
okay, then lets hide it behind a feature flag?

Reply:
let me share little more context :) 

First of all, the the general approach you are taking here is good. This is one of issues that sync events can fix as React Native will no longer need to do the round trip from the main thread to JS thread and back to main thread, resulting in unpleasant flicker.

As the method name suggests, the sync events are still experimental. We are not sure what the right API is at the moment. We are focusing on making the new architecture the default before fully embracing the new capabilities that is potentially breaking. experimental_flushSync works great in isolated cases but might lead to dropped frames when the JS thread is busy. We currently don't have good tools to help engineers find what the main thread is blocked on. 

Once we have a good design for sync events, it will fix some issues with TextInput and significantly simplify the implementation.